examples/multi-gpu-movielens/Multi-GPU Data Processing and Training.ipynb

```diff
@@ -0,0 +1,897 @@
+    "                col, EMBEDDING_TABLE_SHAPES[col][0]  # Input dimension (vocab size)\n",
+    "            ),\n",
+    "            EMBEDDING_TABLE_SHAPES[col][1],  # Embedding output dimension\n",
+    "        )\n",
+    "    )\n",
+    "emb_layer = layers.DenseFeatures(emb_layers)\n",
+    "x_emb_output = emb_layer(inputs)\n",
+    "x = tf.keras.layers.Dense(128, activation=\"relu\")(x_emb_output)\n",
+    "x = tf.keras.layers.Dense(128, activation=\"relu\")(x)\n",
+    "x = tf.keras.layers.Dense(128, activation=\"relu\")(x)\n",
+    "x = tf.keras.layers.Dense(1, activation=\"sigmoid\")(x)\n",
+    "model = tf.keras.Model(inputs=inputs, outputs=x)\n",
+    "loss = tf.losses.BinaryCrossentropy()\n",
+    "opt = tf.keras.optimizers.SGD(0.01 * hvd.size())\n",
+    "opt = hvd.DistributedOptimizer(opt)\n",
+    "checkpoint_dir = \"./checkpoints\"\n",
+    "checkpoint = tf.train.Checkpoint(model=model, optimizer=opt)\n",
+    "\n",
+    "\n",
+    "@tf.function(experimental_relax_shapes=True)\n",
+    "def training_step(examples, labels, first_batch):\n",
+    "    with tf.GradientTape() as tape:\n",
+    "        probs = model(examples, training=True)\n",
+    "        loss_value = loss(labels, probs)\n",
+    "    # Horovod: add Horovod Distributed GradientTape.\n",
+    "    tape = hvd.DistributedGradientTape(tape, sparse_as_dense=True)\n",
+    "    grads = tape.gradient(loss_value, model.trainable_variables)\n",
+    "    opt.apply_gradients(zip(grads, model.trainable_variables))\n",
+    "    # Horovod: broadcast initial variable states from rank 0 to all other processes.\n",
+    "    # This is necessary to ensure consistent initialization of all workers when\n",
+    "    # training is started with random weights or restored from a checkpoint.\n",
+    "    #\n",
+    "    # Note: broadcast should be done after the first gradient step to ensure optimizer\n",
+    "    # initialization.\n",
+    "    if first_batch:\n",
+    "        hvd.broadcast_variables(model.variables, root_rank=0)\n",
+    "        hvd.broadcast_variables(opt.variables(), root_rank=0)\n",
+    "    return loss_value\n",
+    "\n",
+    "\n",
+    "# Horovod: adjust number of steps based on number of GPUs.\n",
+    "for batch, (examples, labels) in enumerate(train_dataset_tf):\n",
+    "    loss_value = training_step(examples, labels, batch == 0)\n",
+    "    if batch % 100 == 0 and hvd.local_rank() == 0:\n",
+    "        print(\"Step #%d\\tLoss: %.6f\" % (batch, loss_value))\n",
+    "hvd.join()\n",
+    "\n",
+    "# Horovod: save checkpoints only on worker 0 to prevent other workers from\n",
+    "# corrupting it.\n",
+    "if hvd.rank() == 0:\n",
+    "    checkpoint.save(checkpoint_dir)\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "traditional-kingdom",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "!horovodrun -np 3 sh hvd_wrapper.sh python trainer.py --dir_in $BASE_DIR --b_size 1024"
```

Comment:
Going to need to change this now to --batch_size

Reply:
I would definitely have forgotten that. Thanks!

Reply:
@jperez999 @karlhigley you also need to change this line in the notebook where it writes the code in a trainer.py:
`BATCH_SIZE = int(args.b_size) or 16384  # Batch Size` It is still `args.b_size`.